File: src/main/java/eotree/EOObject.java
package eotree;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class EOObject extends EOExpr {
    List<EOBndName>     freeAttrs;
    Optional<EOBndName> varargAttr;
    List<EOBndExpr>     bndAttrs;
    Optional<String>    name;

    public EOObject(List<EOBndName> freeAttrs,
                    Optional<EOBndName> varargAttr,
                    List<EOBndExpr> bndAttrs,
                    Optional<String> name) {
        this.freeAttrs = freeAttrs;
        this.varargAttr = varargAttr;
        this.bndAttrs = bndAttrs;
        this.name = name;
    }

    @Override
    public String generateEO(int indent) {
        return "[" +
               freeAttrs.stream()
                       .map(attr -> attr.generateEO(indent))
                       .collect(Collectors.joining(" ")) +
               (varargAttr
                       .map(attr -> " " + attr.generateEO(indent) + "...")
                       .orElse("")) +
               "]" +
               (name
                       .map(n -> " > " + name)
                       .orElse("")
               ) + "\n" +
               bndAttrs.stream()
                       .map(attr -> attr.generateEO(indent + 1))
                       .collect(Collectors.joining("\n"));
    }
}

File: src/main/java/eotree/EOObject.kt
package eotree

import arrow.core.Option
import arrow.core.getOrElse
import java.util.*
import java.util.stream.Collectors

class EOObject(
    var freeAttrs: List<String>,
    var varargAttr: Option<String>,
    var bndAttrs: List<EOBndExpr>,
    var comment: String = "",
) : EOExpr() {
    override fun generateEO(indent: Int): String {
        return listOfNotNull(
            comment.let {
                if (it.isNotEmpty())
                    indent(indent) + it + "\n"
                else
                    null
            },
            indent(indent),
            "[",
            freeAttrs.joinToString(" "),
            varargAttr
                .map { attr -> (if (freeAttrs.isNotEmpty()) " " else "") + attr + "..." }
                .getOrElse { null },
            "]",
            bndAttrs
                .map { attr: EOBndExpr -> "\n" + attr.generateEO(indent + 1) }
                .joinToString("")
        ).joinToString("")
    }

    override fun toString(): String = "[Object]"
}